### PR TITLE
Update `sov-rollup-starter` with `RollupTemplate`

### PR DESCRIPTION
**Pull request by bkolad**
# Description
This PR updates `sov-rollup-starter` with the new `RollupTemplate` which simplifies rollup creation.
The executable `Readme` will be provided in the next PR.

## Testing
Added tests in `sov-rollup-starter`

## Docs
Docs will be provided later.

### CODE REVIEW

File: sov-rollup-starter/src/bin/node.rs [hunk 1]
```diff
@@ -1,53 +1,59 @@
...
-
-    // Read the rollup config from a file
-    let rollup_config_path = env::args()
-        .nth(1)
-        .map(PathBuf::from)
-        .unwrap_or_else(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("rollup_config.toml"));
+    // Initializing logging
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(EnvFilter::from_str("debug,hyper=info").unwrap())
```
Comment (vlopes11) [inline sov-rollup-starter/src/bin/node.rs]: nit: we should probably take this from `Args`

File: sov-rollup-starter/src/bin/node.rs [hunk 2]
```diff
@@ -1,53 +1,59 @@
...
+        .init();
+
+    let args = Args::parse();
+    let rollup_config_path = args.rollup_config_path.as_str();
+
+    let rollup = new_rollup_with_mock_da(
+        &GenesisPaths::from_dir("test-data/genesis/"),
```
Comment (vlopes11) [inline sov-rollup-starter/src/bin/node.rs]: Looks like this should come from `Args` too, with default `test-data/genesis`

File: sov-rollup-starter/src/register_sequencer.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/// register sequencer rpc methods.
+pub(crate) fn register_sequencer<Da>(
+    storage: &<DefaultContext as Spec>::Storage,
+    da_service: Da,
+    methods: &mut jsonrpsee::RpcModule<()>,
+) -> Result<(), anyhow::Error>
+where
+    Da: DaService,
+{
+    let batch_builder = create_batch_builder::<<Da as DaService>::Spec>(storage.clone());
```
Comment (vlopes11) [inline sov-rollup-starter/src/register_sequencer.rs:20]: Just to give context, is it ok to clone here? Meaning: is the storage always expected to be `Arc`-like?

File: sov-rollup-starter/src/register_sequencer.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    let sequencer_rpc = get_sequencer_rpc(batch_builder, da_service);
+    methods
+        .merge(sequencer_rpc)
+        .context("Failed to merge Txs RPC modules")
+}
+
+fn create_batch_builder<Da: DaSpec>(
+    storage: ProverStorage<sov_state::DefaultStorageSpec>,
+) -> FiFoStrictBatchBuilder<Runtime<DefaultContext, Da>, DefaultContext> {
+    let batch_size_bytes = 1024 * 100; // 100 KB
```
Comment (vlopes11) [inline sov-rollup-starter/src/register_sequencer.rs]: Consider adding a comment on the rationale for `100Kb`

File: sov-rollup-starter/src/rollup.rs [hunk 5]
```diff
@@ -1,110 +1,99 @@
...
+use crate::register_sequencer::register_sequencer;
+
+/// Rollup with MockDa.
+pub struct StarterRollup {}
+
+#[async_trait]
+impl RollupTemplate for StarterRollup {
+    type DaService = MockDaService;
+    type GenesisPaths = GenesisPaths<PathBuf>;
+    type Vm = Risc0Host<'static>;
```
Comment (vlopes11) [inline sov-rollup-starter/src/rollup.rs:30]: Any reason we are coupling it to `Risc0`? Might be worthy to comment here

File: sov-rollup-starter/src/rollup.rs [hunk 6]
```diff
@@ -1,110 +1,99 @@
...
-        runner.run_in_process().await?;
-
-        Ok(())
+    fn create_native_storage(
+        &self,
+        rollup_config: &RollupConfig<Self::DaConfig>,
+    ) -> <Self::NativeContext as Spec>::Storage {
+        let storage_config = StorageConfig {
+            path: rollup_config.storage.path.clone(),
+        };
```
Comment (vlopes11) [inline sov-rollup-starter/src/rollup.rs:78]: nit: it looks like `StorageConfig` should implement `From<RollupConfig<T>>`

File: sov-rollup-starter/src/rollup.rs [hunk 7]
```diff
@@ -1,110 +1,99 @@
...
-        )?;
-
-        runner.start_rpc_server(rpc_module, channel).await;
-        runner.run_in_process().await?;
-
-        Ok(())
+    fn create_native_storage(
+        &self,
+        rollup_config: &RollupConfig<Self::DaConfig>,
+    ) -> <Self::NativeContext as Spec>::Storage {
```
Comment (vlopes11) [inline sov-rollup-starter/src/rollup.rs]: This function handles I/O and should probably be fallible

File: sov-rollup-starter/src/rollup.rs [hunk 8]
```diff
@@ -1,110 +1,99 @@
...
-    /// exposed by the node.
-    fn rpc_module(
-        &mut self,
-        storage: <DefaultContext as Spec>::Storage,
-    ) -> anyhow::Result<RpcModule<()>> {
-        let mut module = get_rpc_methods::<DefaultContext, Da::Spec>(storage.clone());
+    fn create_rpc_methods(
+        &self,
+        storage: &<Self::NativeContext as Spec>::Storage,
+        ledger_db: &LedgerDB,
```
Comment (vlopes11) [inline sov-rollup-starter/src/rollup.rs:85]: Shouldn't the `LedgerDB` also be generic for the trait?

**Review by vlopes11 — COMMENTED** [review]
Looks good overall! Added a couple of preliminary questions for the definitive review

File: sov-rollup-starter/src/register_sequencer.rs (hunk 3 above)
Comment (bkolad) [reply sov-rollup-starter/src/register_sequencer.rs:20]: Yes the storage is behind an `Arc`

File: sov-rollup-starter/src/rollup.rs (hunk 5 above)
Comment (bkolad) [reply sov-rollup-starter/src/rollup.rs:30]: Yes, we have `Rollup` per `DA` and `VM `so the `DaService` & `Vm` needs to be hardcoded at this point.

File: sov-rollup-starter/src/rollup.rs (hunk 8 above)
Comment (bkolad) [reply sov-rollup-starter/src/rollup.rs:85]: Probably yes. The thing is that the whole code depends on the concrete type, and we didn't need to abstract it so far.  So I would propose to do it when we see benefits.

File: sov-rollup-starter/src/rollup.rs (hunk 7 above)
Comment (bkolad) [reply sov-rollup-starter/src/rollup.rs]: done

File: sov-rollup-starter/src/bin/node.rs (hunk 1 above)
Comment (bkolad) [reply sov-rollup-starter/src/bin/node.rs]: I changed it to `EnvFilter::from_default_env()` so now we read form  `env` variable.

File: sov-rollup-starter/src/bin/node.rs (hunk 2 above)
Comment (bkolad) [reply sov-rollup-starter/src/bin/node.rs]: done

File: sov-rollup-starter/src/register_sequencer.rs (hunk 4 above)
Comment (bkolad) [reply sov-rollup-starter/src/register_sequencer.rs]: fixed

File: sov-rollup-starter/src/rollup.rs (hunk 6 above)
Comment (bkolad) [reply sov-rollup-starter/src/rollup.rs:78]: I think `StorageConfig` will need to be part of `RollupConfig`. This will be fixed with  #1083.

**Review by vlopes11 — APPROVED** [review]
Looks good!

**Pull request merged**